src/world/action/ultra_boots.c

```diff
@@ -1,5 +1,200 @@
+        playerStatus->gravityIntegrator[2] = 3.44694f;
+        playerStatus->gravityIntegrator[3] = -0.75f;
+        suggest_player_anim_setUnkFlag(0x80000);
+        disable_player_input();
+        playerStatus->flags |= 0x200;
+        gCameras->moveFlags = gCameras->moveFlags | 1;
+        temp_f0 = clamp_angle(playerStatus->targetYaw - gCameras[gCurrentCameraID].currentYaw);
+        phi_f4 = -60.0f;
+        if (temp_f0 <= 180.0f) {
+            phi_f4 = 60.0f;
+        }
+        playerStatus->unk_D4 = phi_f4;
+    }
+    if (playerStatus->fallState < 4) {
+        if (playerStatus->unk_D4 >= 0.0f) {
+            playerStatus->spriteFacingAngle += playerStatus->unk_D4;
+            if (playerStatus->spriteFacingAngle >= 360.0f) {
+                playerStatus->spriteFacingAngle -= 360.0f;
+            }
+        } else {
+            playerStatus->spriteFacingAngle += playerStatus->unk_D4;
+            if (playerStatus->spriteFacingAngle < 0.0f) {
+                playerStatus->spriteFacingAngle += 360.0f;
+            }
+        }
+    }
+    if (playerStatus->gravityIntegrator[0] <= 0.0f) {
+        *(&gSpinHistoryPosY + (gSpinHistoryBufferPos)) = playerStatus->position.y;
+    }
+
+    *(&gSpinHistoryPosAngle + gSpinHistoryBufferPos++) = playerStatus->spriteFacingAngle;
+
+    if (gSpinHistoryBufferPos >= 6) {
+        gSpinHistoryBufferPos = 0;
+    }
+
+    switch (playerStatus->fallState) {
+    case 0:
+        temp_f20 = integrate_gravity();
+        playerStatus->position.y = player_check_collision_below(temp_f20, &sp10);
+        if (sp10 >= 0 && collisionStatus->currentFloor & 0x4000 && (get_entity_type(collisionStatus->currentFloor) - 7 < 2)) {
+            currentFloorEntity = get_entity_by_index(collisionStatus->currentFloor);
+            currentFloorEntity->collisionFlags |= 1;
+            disable_player_input();
+            playerStatus->fallState = 0xB;
+        } else {
+            if (temp_f20 <= 0.0f) {
+                record_jump_apex();
+                playerStatus->framesOnGround = 3;
+                playerStatus->flags |= 4;
+                playerStatus->fallState++;
+                sfx_play_sound_at_player(0x147, 0);
+            }
+            if (sp10 >= 0) {
+                playerStatus->flags &= ~0x00020008;
+                set_action_state(ACTION_STATE_LAND);
+            }
+        }
+        break;
+    case 1:
+        if (--playerStatus->framesOnGround <= 0) {
+            playerStatus->fallState++;
+        }
+        break;
+    case 2:
+        temp_f20 = integrate_gravity();
+        playerStatus->position.y = player_check_collision_below(temp_f20, &sp10);
+        if (temp_f20 < -100.0f) {
+            playerStatus->gravityIntegrator[3] = 0.0f;
+            playerStatus->gravityIntegrator[2] = 0.0f;
+            playerStatus->gravityIntegrator[1] = 0.0f;
+            playerStatus->gravityIntegrator[0] = -100.0f;
+        }
+        if (sp10 >= 0) {
```

Comment:
The control flow from here on is a bit hard to figure out with all the early returns. I think you might be able to if/else a lot of this